### DISCUSSION TITLE
How To Split Multiple Authors Into List Items In Properties

### DISCUSSION THREAD

**Discussion by tillmanj** [aidenlx/obsidian-zotero#260] [Help]
In my zt-note.eta template I use the following to list N authors (or editors) on a single line as links to files.

`<% if (it.authors && it.authors[0]) { %><% it.authors.forEach(function(author){ %>   [[<%= author %>]] <% }) %><% } else if (it.creators && it.creators[0]) { %><%_ it.creators.forEach(function(creator){ %>   [[<%= creator %>]] <% }) %> *(editor)*<% } %>`

However, I also store the author names in a Property (FKA Frontmatter) titled 'authors'. However, adapting the above code to work with the property (type=list) is proving difficult.

Things I have tried:

`<% if (it.authors && it.authors[0]) { %><% it.authors.forEach(function(author){ %> <%= author %> <% }) %><% } else if (it.creators && it.creators[0]) { %><%_ it.creators.forEach(function(creator){ %> <%= creator %> <% }) %> <% } %>`
This list all the authors as a single tag

`<% if (it.authors && it.authors[0]) { %><% it.authors.forEach(function(author){ %> [<%= author %>] <% }) %><% } else if (it.creators && it.creators[0]) { %><%_ it.creators.forEach(function(creator){ %> [<%= creator %>] <% }) %> <% } %>`
This fails with "Failed to render YAML"

`<% if (it.authors && it.authors[0]) { %><% it.authors.forEach(function(author){ %>   <%= author + '\n' %> <% }) %><% } else if (it.creators && it.creators[0]) { %><%_ it.creators.forEach(function(creator){ %>   <%= creator + '\n' %> <% }) %> <% } %>`
Also fails with "Failed to render YAML"

Any ideas how I can go about inserting each author as a single entry in a list type Property?

**Comment by Arc13Tangent** [comment]
Hi, I use the following:
```
authors: <% for (const $it of it.creators) { %> 
  - <%= $it.firstName[0] %>. <%= $it.lastName %>
<% } %>
```
This is the result:
![Screen Shot 2024-08-30 at 21 36 10](https://github.com/user-attachments/assets/f8240b41-9fa3-4dd2-8264-c0bcba487fde)

not entirely sure if it achieves the effect you're looking for 😆
(Thank you for bringing up this question. I struggled with this problem before and found this thread, but I eventually got the desired result somewhat by randomly trying different things XD)

**Comment by dcgk** [comment]
Hi, here is my snippet for authors. I hope this helps.

The logic is this:
- If authors are available, we list authors. Otherwise we list creators.
- We list each author/creator as `lastName, firstName`
- If lastName is not available we fall back on `fullName`
- If there are creators who have not already been listed as authors, we also list them.
- If creators have a creatorType that is not "author" , then the creatorType is appended to the name e.g. *(editor)*

The output is a correctly formated YAML list: a new line beginning with `-` for each author/creator 

```
<%
  // Processing logic for authors, setup functions and variables
  // Function to format a name entry based on available name fields
  function formatName(entry) {
    if (entry.lastName && entry.firstName) {
      return `${entry.lastName}, ${entry.firstName}`;
    } else if (entry.fullname) {
      return entry.fullname;
    } else {
      return "Unknown"; // Fallback in case no name fields are present
    }
  }

  // Function to format the entire entry and append the type if not an "author"
  function formatEntry(entry, typeCheck = false) {
    let formattedName = formatName(entry);
    if (typeCheck && entry.type !== "author") {
      formattedName += ` *(${entry.type})*`;
    }
    return formattedName;
  }

  const authorIDs = new Set(); // To store itemIDs of authors
  let authorOutput = ''; // Store formatted authors for use later

  // Process authors
  if (Array.isArray(it.authors) && it.authors.length > 0) {
    it.authors.forEach(function(author) {
      const authorFormatted = formatEntry(author);
      if (author.itemID) {
        authorIDs.add(author.itemID); // Store author itemID to check later
      }
      authorOutput += `- ${authorFormatted}\n`;
    });
  }

  // Process creators, avoid duplication of authors
  if (Array.isArray(it.creators) && it.creators.length > 0) {
    it.creators.forEach(function(creator) {
      if (!authorIDs.has(creator.itemID)) { // Check if creator itemID has not appeared in authors
        const creatorFormatted = formatEntry(creator, true);
        authorOutput += `- ${creatorFormatted}\n`;
      }
    });
  }
%>
author:
<%= authorOutput %>
```

**Comment by aidenlx** [comment]
This moved out of the template body and into frontmatter fields in v2, which removes the hand-formatting entirely.

At **Settings > Templates > Frontmatter**, add a field with key `authors` and a Liquid expression:

```
zt.creators | map: "fullName"
```

The expression returns a real array and ZotLit serializes it to a YAML list, so there are no `- ` lines to emit by hand and nothing to escape — which is what was producing *Failed to render YAML* in your attempts.

Creator objects expose `family`, `given`, `literal`, `role` and `fullName`, so `map: "family"` or a filter on `role` works the same way. Setting the merge strategy to **Append arrays** keeps any names you add manually.

Reference: https://zotlit.aidenlx.site/docs/reference/templates/frontmatter

v2 beta is available now: https://zotlit.aidenlx.site/docs/install-zotlit
Migration guide: https://zotlit.aidenlx.site/docs/how-to/migrate-from-v1
